cpp/src/fil/infer.cu

```diff
@@ -452,6 +459,126 @@ struct tree_aggregator_t<NITEMS, GROVE_PER_CLASS_MANY_CLASSES> {
   }
 };
 
+template <int NITEMS>
+struct tree_aggregator_t<NITEMS, VECTOR_LEAF> {
+  vec<NITEMS, float>* per_class_margin;
+  vec<NITEMS, int>* vector_leaf_indices;
+  int* thread_num_rows;
+  int num_classes;
+  int num_threads_per_class;
+  float* vector_leaf;
+
+  static size_t smem_finalize_footprint(size_t data_row_size, int num_classes,
+                                        bool predict_proba) {
+    return 0;
+  }
+  static size_t smem_accumulate_footprint(int num_classes) {
+    return sizeof(vec<NITEMS, float>) * num_classes *
+             max(1, FIL_TPB / num_classes) +
+           sizeof(vec<NITEMS, int>) * FIL_TPB + sizeof(int) * FIL_TPB;
+  }
+
+  __device__ __forceinline__ tree_aggregator_t(predict_params params,
+                                               void* accumulate_workspace,
+                                               void* finalize_workspace,
+                                               float* vector_leaf)
+    : num_classes(params.num_classes),
+      num_threads_per_class(max(1, blockDim.x / params.num_classes)),
+      vector_leaf(vector_leaf) {
+    // Assign workspace
+    char* ptr = (char*)accumulate_workspace;
+    per_class_margin = (vec<NITEMS, float>*)ptr;
+    ptr += sizeof(vec<NITEMS, float>) * num_classes * num_threads_per_class;
+    vector_leaf_indices = (vec<NITEMS, int>*)ptr;
+    ptr += sizeof(vec<NITEMS, int>) * blockDim.x;
+    thread_num_rows = (int*)ptr;
+
+    // Initialise shared memory
+    for (int i = threadIdx.x; i < num_classes * num_threads_per_class;
+         i += blockDim.x) {
+      per_class_margin[i] = vec<NITEMS, float>();
+    }
+    vector_leaf_indices[threadIdx.x] = vec<NITEMS, int>();
+    thread_num_rows[threadIdx.x] = 0;
+    // __syncthreads() is called in infer_k
+  }
+  __device__ __forceinline__ void accumulate(
+    vec<NITEMS, int> single_tree_prediction, int tree, int num_rows) {
+    // Perform a transpose in shared memory
+    // Assign each thread to a class, so they can accumulate without atomics
+    __syncthreads();
+    // Write indices to shared memory
+    vector_leaf_indices[threadIdx.x] = single_tree_prediction;
+    thread_num_rows[threadIdx.x] = num_rows;
+    __syncthreads();
+    for (int i = threadIdx.x; i < num_classes * num_threads_per_class;
```

Comment:
Could you add more documentation to this nest of loops explaining what you're trying to do? Explanations of what each loop does when there are few classes and very many classes would also be useful.

Reply:
I've expanded this out with some comments.

Reply:
now it makes sense, thank you!